Source/WebCore/dom/InternalObserver.h

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (C) 2024 Keith Cirkel <webkit@keithcirkel.co.uk>. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "ActiveDOMObject.h"
+#include "SubscriptionObserver.h"
+#include "JSSubscriptionObserverCallback.h"
+#include <wtf/RefCounted.h>
+
+namespace WebCore {
+
+class InternalObserver : public ActiveDOMObject, public RefCounted<InternalObserver> {
+public:
+    virtual ~InternalObserver() = default;
+
+    virtual void next(JSC::JSValue value) = 0;
+    virtual void complete()
+    {
+        m_active = false;
+    }
+
+    virtual void error(JSC::JSValue value)
+    {
+        auto* context = scriptExecutionContext();
+        if (!context)
+            return;
+
+        auto* globalObject = context->globalObject();
+        if (!globalObject)
+            return;
+
+        Ref vm = globalObject->vm();
+        JSC::JSLockHolder lock(vm);
+
+        reportException(globalObject, JSC::Exception::create(vm, value));
+    }
+
+    // ActiveDOMObject
+    void ref() const final { RefCounted::ref(); }
+    void deref() const final { RefCounted::deref(); }
+
+protected:
+    bool m_active = true;
+
+    InternalObserver(ScriptExecutionContext& context)
+        : ActiveDOMObject(&context)
+    {
+        suspendIfNeeded();
+    }
+
+    // ActiveDOMObject
+    void stop() override {}
+    bool virtualHasPendingActivity() const override { return m_active; }
+};
+
+
+class InternalObserverFromScript final : public InternalObserver {
```

Comment:
This class allows us to simpify `Subscriber` so that it only ever takes `InternalObserver`s, by taking the existing way `Subscriber` consumes a  `SubscriptionObserver` and mapping that to an `InternalObserver` delegate instead. I called this `InternalObserverFromScript` as it's `InternalObserver` that will only be created from the `Observer#subscribe()` IDL.

Reply:
These comments you've  made would be great to add to the commit message. I like to use the changed files section for commentary like this.

e.g..

```
* Source/WebCore/dom/InternalObserver.h: Added
(WebCore::InternalObserver::complete):
(WebCore::InternalObserver::error):
(WebCore::InternalObserver::InternalObserver):
    <my comments here>
```